pkg/nsx/services/nsxserviceaccount/cluster.go

```diff
@@ -0,0 +1,289 @@
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/util"
+)
+
+const (
+	siteId             = "default"
+	enforcementpointId = "default"
+	PortRestAPI        = "rest-api"
+	PortNSXRPCFwdProxy = "nsx-rpc-fwd-proxy"
+	SecretSuffix       = "-nsx-cert"
+	SecretCAName       = "ca.crt"
+	SecretCertName     = "tls.crt"
+	SecretKeyName      = "tls.key"
+)
+
+var (
+	log = logger.Log
+
+	isProtectedTrue = true
+	vpcRole         = "ccp_internal_operator"
+	readerPath      = "/"
+	readerRole      = "cluster_info_reader"
+
+	antreaClusterResourceType = "AntreaClusterControlPlane"
+	revision1                 = int64(1)
+)
+
+type NSXServiceAccountService struct {
+	common.Service
+	PrincipalIdentityStore   *PrincipalIdentityStore
+	ClusterControlPlaneStore *ClusterControlPlaneStore
+}
+
+// InitializeNSXServiceAccount sync NSX resources
+func InitializeNSXServiceAccount(service common.Service) (*NSXServiceAccountService, error) {
+	wg := sync.WaitGroup{}
+	wgDone := make(chan bool)
+	fatalErrors := make(chan error)
+
+	wg.Add(2)
+	nsxServiceAccountService := &NSXServiceAccountService{Service: service}
+
+	nsxServiceAccountService.SetUpStore()
+	go nsxServiceAccountService.InitializeResourceStore(&wg, fatalErrors, common.ResourceTypePrincipalIdentity, nsxServiceAccountService.PrincipalIdentityStore)
+	go nsxServiceAccountService.InitializeResourceStore(&wg, fatalErrors, common.ResourceTypeClusterControlPlane, nsxServiceAccountService.ClusterControlPlaneStore)
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		return nsxServiceAccountService, err
+	}
+
+	return nsxServiceAccountService, nil
+}
+
+func (s *NSXServiceAccountService) SetUpStore() {
+	s.PrincipalIdentityStore = &PrincipalIdentityStore{ResourceStore: common.ResourceStore{
+		Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeNSXServiceAccountCRUID: indexFunc}),
+		BindingType: mpmodel.PrincipalIdentityBindingType(),
+	}}
+	s.ClusterControlPlaneStore = &ClusterControlPlaneStore{ResourceStore: common.ResourceStore{
+		Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeNSXServiceAccountCRUID: indexFunc}),
+		BindingType: model.ClusterControlPlaneBindingType(),
+	}}
+}
+
+func (s *NSXServiceAccountService) CreateOrUpdateNSXServiceAccount(ctx context.Context, obj *v1alpha1.NSXServiceAccount) error {
+	clusterName := s.getClusterName(obj.Namespace, obj.Name)
+	normalizedClusterName := util.NormalizeId(clusterName)
+	// TODO: Support default project name in config as project
+	project := s.NSXConfig.CoeConfig.Cluster
```

Comment:
in this PR https://github.com/vmware-tanzu/nsx-operator/pull/183 will contain Project info.

Reply:
Do we need `s.NSXConfig.CoeConfig.Cluster` as a fallback? Or can we guarantee `nsxt_project` will exist always?

Reply:
Currently for NSXServiceAccount, you do not need a real NSX VPC there, right? you just need a VPCPath right?
If yes, I think you can use cluster s Project for now. With VPC networking supported, projects need to be created by user and set as input when creating namespace.

Reply:
OK. thanks.